### PR TITLE
Skaffold system check

### PR DESCRIPTION
**Pull request by ZariaHoward**
Checks to see if Skaffold is on system when the user tries to Deploy.

### CODE REVIEW

**Comment by etanshaul** [comment]
Some additional tips:

**Meaningful commit message** 
The message should be descriptive of the problem (and potentially the fix as well). See previous commits in the repo for some examples. In this case, a possible message could be `Checks to ensure Skaffold is on the PATH and shows error message if not`

**PR Description** 
Start the PR with `fixes #[bug number here]` on the first line, if there is an applicable issue. GitHub will then close the associated issue automatically when merging into master. 

Add sufficient detail to the PR description to describe your solution. Also, screenshots are always helpful when there are any UI facing changes. In this case, a screenshot of the log message would be helpful.

**Build Checks**
As @ivanporty mentioned, you can run `gradlew check` locally to ensure that all of the checks pass, including unit tests. If they are not passing locally then they will likely not pass our Kokoro CI.

**Unit Tests**
Add unit tests whenever possible. In some cases it may not make sense, but the reasons should be discussed / called out. See some of our existing unit tests in this repo for helpful patterns (for example, we use the `ContainerToolsRule` test rule for our unit tests).

File: skaffold/src/main/kotlin/com/google/container/tools/skaffold/run/SkaffoldRunConfigurations.kt [hunk 1]
```diff
@@ -101,4 +106,18 @@ abstract class AbstractSkaffoldRunConfiguration(
 
         XmlSerializer.serializeInto(this, element)
     }
+
+    @Throws(RuntimeConfigurationException::class)
```
Comment (etanshaul) [inline skaffold/src/main/kotlin/com/google/container/tools/skaffold/run/SkaffoldRunConfigurations.kt]: Probably don't want to commit these changes?

File: skaffold/src/main/kotlin/com/google/container/tools/skaffold/SkaffoldExecutorService.kt [hunk 2]
```diff
@@ -85,6 +85,8 @@ abstract class SkaffoldExecutorService {
         }
 
         try {
+            //this is where that whole no skaffold directory thing is happening
+            // Wait why isnt going into the catch
```
Comment (etanshaul) [inline skaffold/src/main/kotlin/com/google/container/tools/skaffold/SkaffoldExecutorService.kt]: oh it seems like this may not be ready yet for review? if so, its fine to put out the PR, just don't assign it for review until it is ready.

File: skaffold/src/main/kotlin/com/google/container/tools/skaffold/run/SkaffoldCommandLineState.kt [hunk 3]
```diff
@@ -60,6 +64,24 @@ class SkaffoldCommandLineState(
             throw ExecutionException(message("skaffold.no.file.selected.error"))
         }
 
+        //call skaffold version --- look at createProcess function, the command list lets you know if skaffold isn't on the system by process of elimination
+        if (Paths.get("skaffold").exists()) {
```
Comment (etanshaul) [inline skaffold/src/main/kotlin/com/google/container/tools/skaffold/run/SkaffoldCommandLineState.kt]: I think a better way to check that Skaffold is on the PATH is to check the system path directly and see if `skaffold` is there. Its generally advised not to use exception control flow in this way. Something like this (with comments added for clarity):

```
fun isExecutableInPath(executable: String): Boolean {
   val path = System.getenv("PATH") // get the full path string
   return path.split(File.pathSeparator) // split the path string by the path separator
      .asSequence() // use sequence for lazy evaluation
      .map { it + File.separator + executable } // convert each to a possible path to the executable
      .any {
         File(it).exists() // does the file exist?
      }
}
```

then calling `isExecutableInPath("skaffold")` should do the check

File: skaffold/src/main/kotlin/com/google/container/tools/skaffold/run/SkaffoldCommandLineState.kt [hunk 4]
```diff
@@ -60,6 +64,24 @@ class SkaffoldCommandLineState(
...
+            val testSkaffoldCommand = Paths.get("skaffold").toString() + " " + (executionMode.modeFlag)
```
Comment (etanshaul) [inline skaffold/src/main/kotlin/com/google/container/tools/skaffold/run/SkaffoldCommandLineState.kt]: one reason the build is failing is because this line is over 100 chars. you can see this by running `./gradlew check`. these currently have to be fixed manually because `./gradlew spotlessApply` does not yet fix line length violations

Comment (etanshaul) [reply skaffold/src/main/kotlin/com/google/container/tools/skaffold/run/SkaffoldCommandLineState.kt]: by "manually" I mean you can run a reformat in intellij: code -> reformat code

File: skaffold/src/main/kotlin/com/google/container/tools/skaffold/run/SkaffoldRunConfigurations.kt [hunk 5]
```diff
@@ -21,14 +21,19 @@ import com.google.container.tools.skaffold.run.ui.SkaffoldDevSettingsEditor
 import com.google.container.tools.skaffold.run.ui.SkaffoldSingleRunSettingsEditor
 import com.intellij.execution.Executor
 import com.intellij.execution.configurations.ConfigurationFactory
+import com.intellij.execution.configurations.ConfigurationPerRunnerSettings
```
Comment (etanshaul) [inline skaffold/src/main/kotlin/com/google/container/tools/skaffold/run/SkaffoldRunConfigurations.kt]: Some of these imports are unused also causing the build to fail. an easy way to fix this is to run "optimize imports":

`Code -> Optimize Imports`

Comment (etanshaul) [reply skaffold/src/main/kotlin/com/google/container/tools/skaffold/run/SkaffoldRunConfigurations.kt]: `./gradlew spotlessApply` should also fix that _I think_

File: skaffold/src/test/kotlin/com/google/container/tools/skaffold/run/SkaffoldRunConfigurationsTest [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+import io.mockk.mockkStatic
+import io.mockk.slot
+import org.jdom.Element
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+import java.io.File
+
+/** Unit tests for [SkaffoldCommandLineState] */
+class SkaffoldCommandLineStateTest {
```
Comment (etanshaul) [inline skaffold/src/test/kotlin/com/google/container/tools/skaffold/run/SkaffoldRunConfigurationsTest]: the class name should typically match the file name. did you mean to call this class `SkaffoldRunConfigurationsTest`?

Comment (etanshaul) [reply skaffold/src/test/kotlin/com/google/container/tools/skaffold/run/SkaffoldRunConfigurationsTest]: also it looks like you added it as a text file instead of a `.kt` file. 

When creating a new Kotlin file in the IDE, use:
`right click on the folder` -> new -> Kotline File/class

File: skaffold/src/main/kotlin/com/google/container/tools/skaffold/SkaffoldExecutorService.kt [hunk 7]
```diff
@@ -43,7 +43,7 @@ abstract class SkaffoldExecutorService {
     }
 
     /** Path for Skaffold executable, any form supported by [ProcessBuilder] */
-    protected abstract var skaffoldExecutablePath: Path
+    open var skaffoldExecutablePath: Path = Paths.get("skaffold")
```
Comment (etanshaul) [inline skaffold/src/main/kotlin/com/google/container/tools/skaffold/SkaffoldExecutorService.kt]: this change shouldn't be necessary. the `DefaultSkaffoldExecutorService` is overriding this and setting the value

File: skaffold/src/main/kotlin/com/google/container/tools/skaffold/run/SkaffoldCommandLineState.kt [hunk 8]
```diff
@@ -60,6 +64,24 @@ class SkaffoldCommandLineState(
...
+            try {
+                SkaffoldProcess(
+                    SkaffoldExecutorService.instance.createProcess(
+                        File(projectBaseDir.path),
+                        listOf(testSkaffoldCommand)
+                    ),
+                    commandLine = testSkaffoldCommand//skaffold executable path
+                )
+            }catch (e: Exception){
+                throw ExecutionException("skaffold path in system is incorrect")
```
Comment (etanshaul) [inline skaffold/src/main/kotlin/com/google/container/tools/skaffold/run/SkaffoldCommandLineState.kt]: any time we have UI visible messages in the code, we should put them in the message bundle. see line 64 for an existing example of this:

`throw ExecutionException(message("skaffold.no.file.selected.error"))`

File: skaffold/src/main/kotlin/com/google/container/tools/skaffold/run/SkaffoldRunConfigurations.kt [hunk 9]
```diff
@@ -106,18 +111,33 @@ abstract class AbstractSkaffoldRunConfiguration(
...
-        throw RuntimeConfigurationException("Hitting function2222")
     }
 
 
-    override fun checkSettingsBeforeRun(){
-        throw RuntimeConfigurationException("Hitting function")
+    companion object {
+        var skaffoldPath: Path? = Paths.get("skaffold")
+        @JvmStatic
+        fun checkConfiguration(project:Project) {
```
Comment (etanshaul) [inline skaffold/src/main/kotlin/com/google/container/tools/skaffold/run/SkaffoldRunConfigurations.kt]: can this just live in the class itself instead of putting it in a companion object?

File: skaffold/src/main/kotlin/com/google/container/tools/skaffold/run/SkaffoldRunConfigurations.kt [hunk 10]
```diff
@@ -106,18 +111,33 @@ abstract class AbstractSkaffoldRunConfiguration(
...
+            val projectBaseDir: VirtualFile? = project.guessProjectDir()
+            if (skaffoldPath?.exists() == true && projectBaseDir != null ) {
+                val testSkaffoldCommand = skaffoldPath.toString() + " " + (SkaffoldExecutorSettings.ExecutionMode.SINGLE_RUN.modeFlag)
```
Comment (etanshaul) [inline skaffold/src/main/kotlin/com/google/container/tools/skaffold/run/SkaffoldRunConfigurations.kt]: don't forget to update this to either do a version check. or update the path checking strategy using the code snippet suggestion I gave in another comment

**Comment by etanshaul** [comment]
@ivanporty for review as well

File: skaffold/src/main/kotlin/com/google/container/tools/skaffold/run/SkaffoldRunConfigurations.kt (hunk 10 above)
Comment (etanshaul) [reply skaffold/src/main/kotlin/com/google/container/tools/skaffold/run/SkaffoldRunConfigurations.kt]: one option to help with testing:

move the path check to `SkaffoldFileService`. This is an applicationService defined in `skaffold.xml`. You can swap these out in unit tests with a mock using the `@TestService` annotation. I can walk you through this.

File: skaffold/src/main/kotlin/com/google/container/tools/skaffold/run/SkaffoldRunConfigurations.kt (hunk 9 above)
Comment (ZariaHoward) [reply skaffold/src/main/kotlin/com/google/container/tools/skaffold/run/SkaffoldRunConfigurations.kt]: Apparently Kotlin doesn't have static members of abstract classes from what I read? I wanted the function to be able to be called without creating an instance, an according to the docs that is where companion objects come in.

Comment (etanshaul) [reply skaffold/src/main/kotlin/com/google/container/tools/skaffold/run/SkaffoldRunConfigurations.kt]: makes sense.

see my other suggestion about moving this utility to `SkaffoldFileService`. This should help better structure things (since you are using this check in several contexts); as well as help make unit testing easier.

File: skaffold/src/main/kotlin/com/google/container/tools/skaffold/run/SkaffoldRunConfigurations.kt (hunk 10 above)
Comment (etanshaul) [reply skaffold/src/main/kotlin/com/google/container/tools/skaffold/run/SkaffoldRunConfigurations.kt]: sorry - I think `SkaffoldExecutorService` is the better place for this

File: skaffold/src/main/kotlin/com/google/container/tools/skaffold/run/SkaffoldRunConfigurations.kt [hunk 11]
```diff
@@ -101,4 +111,33 @@ abstract class AbstractSkaffoldRunConfiguration(
...
         XmlSerializer.serializeInto(this, element)
     }
+    override fun checkConfiguration(){
+        checkConfiguration(project)
+
+    }
+
+
+    companion object {
+        var skaffoldPath: Path? = Paths.get("skaffold")
```
Comment (ivanporty) [inline skaffold/src/main/kotlin/com/google/container/tools/skaffold/run/SkaffoldRunConfigurations.kt]: We probably don't need this at all - at the end we call `SkaffoldExecutorService.instance.createProcess` and if it works, it should work for general case as well.

File: skaffold/src/main/kotlin/com/google/container/tools/skaffold/run/SkaffoldRunConfigurations.kt (hunk 10 above)
Comment (ivanporty) [reply skaffold/src/main/kotlin/com/google/container/tools/skaffold/run/SkaffoldRunConfigurations.kt]: +1 on relocating this as a normal member function to `SkaffoldExecutorService`

File: skaffold/src/main/kotlin/com/google/container/tools/skaffold/run/SkaffoldCommandLineState.kt [hunk 12]
```diff
@@ -60,6 +62,12 @@ class SkaffoldCommandLineState(
             throw ExecutionException(message("skaffold.no.file.selected.error"))
         }
 
+        try {
+            AbstractSkaffoldRunConfiguration.checkConfiguration(environment.project)
```
Comment (ivanporty) [inline skaffold/src/main/kotlin/com/google/container/tools/skaffold/run/SkaffoldCommandLineState.kt]: So if we put this into `SkaffoldExecutorService`, we have the following `SkaffoldExecutorService.instance.checkConfiguration`, with the added benefit of easier mocking and dynamic service lookup.

File: skaffold/src/test/kotlin/com/google/container/tools/skaffold/run/SkaffoldRunConfigurationsTest.kt [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+    fun `A runtime warning is not shown if skaffold is in the system PATH`() {
+
+        val testObj = mockk<AbstractSkaffoldRunConfiguration.Companion>()
+
+        every { testObj.skaffoldPath } returns Paths.get("skaffold")
+
+        testObj.checkConfiguration(mockExecutionEnvironment.project)
+
+    }
+}
```
Comment (ivanporty) [inline skaffold/src/test/kotlin/com/google/container/tools/skaffold/run/SkaffoldRunConfigurationsTest.kt]: I think we add a newline at the end of our files - I believe this is a common GitHub practice.

File: skaffold/src/main/kotlin/com/google/container/tools/skaffold/run/ui/BaseSkaffoldSettingsEditor.kt [hunk 14]
```diff
@@ -113,6 +114,10 @@ open class BaseSkaffoldSettingsEditor<T : AbstractSkaffoldRunConfiguration>(
             throw ConfigurationException(message("skaffold.invalid.file.error"))
         }
 
+        if(true){
```
Comment (ivanporty) [inline skaffold/src/main/kotlin/com/google/container/tools/skaffold/run/ui/BaseSkaffoldSettingsEditor.kt]: This is probably a test code, to be removed once the check is in place?

File: skaffold/src/main/kotlin/com/google/container/tools/skaffold/run/SkaffoldCommandLineState.kt (hunk 12 above)
Comment (etanshaul) [reply skaffold/src/main/kotlin/com/google/container/tools/skaffold/run/SkaffoldCommandLineState.kt]: thanks @ivanporty! @ZariaHoward and I discussed doing exactly this

File: skaffold/src/main/kotlin/com/google/container/tools/skaffold/run/SkaffoldCommandLineState.kt (hunk 3 above)
Comment (etanshaul) [reply skaffold/src/main/kotlin/com/google/container/tools/skaffold/run/SkaffoldCommandLineState.kt]: @ivanporty had a great point that this doesn't check that the file is executable. I would suggest ammending the last line:

`File(it).exists() && File(it).canExecute()`

File: skaffold/src/main/resources/messages/SkaffoldBundle.properties [hunk 15]
```diff
@@ -42,3 +42,4 @@ skaffold.override.image.repo.tooltip=<html>The default image repository allows y
 skaffold.no.file.selected.error=Skaffold configuration file is not selected.
 skaffold.invalid.file.error=Skaffold configuration file doesn't exist or invalid.
 skaffold.corrupted.run.settings=Your Skaffold run configuration is corrupted. Please re-create it to fix.
+skaffold.not.on.system.error=Skaffold path in system is incorrect. Install Skaffold. See required dependencies <a href="https://github.com/GoogleContainerTools/google-container-tools-intellij#prerequisites-and-required-dependencies">here</a>:
```
Comment (etanshaul) [inline skaffold/src/main/resources/messages/SkaffoldBundle.properties]: nit suggestion:

`Cloud not detect Skaffold installation. Please install Skaffold. See...`

File: skaffold/src/test/kotlin/com/google/container/tools/skaffold/DefaultSkaffoldExecutorServiceTest.kt [hunk 16]
```diff
@@ -226,4 +227,16 @@ class DefaultSkaffoldExecutorServiceTest {
 
         assertThat(result.commandLine).isEqualTo("skaffold dev --filename test.yaml")
     }
+
+    @Test
+    fun `isSkaffoldAvailable returns true when skaffold is available`() {
+        assertThat(defaultSkaffoldExecutorService.isSkaffoldAvailable()).isTrue()
```
Comment (etanshaul) [inline skaffold/src/test/kotlin/com/google/container/tools/skaffold/DefaultSkaffoldExecutorServiceTest.kt]: this will only work if Skaffold is available on the system running this unit test (will probably fail in CI since skaffold is not installed on those machines). you probably want to mock the path as you do in the test below. You'll also then need to mock a file so that `isSkaffoldAvailable` finds the skaffold executable.

for this you can use the `@TestFile` annotation:

at the class level add the following (which creates a test file named `skaffold` that gets cleaned up automatically when the test completes):
```
@TestFile(name = "skaffold")
private lateinit var testSkaffoldFile: File
```

Then you can mock the system path to return `testSkaffoldFile.parent`. This simulates having the test file's directory on the system path.

File: skaffold/src/test/kotlin/com/google/container/tools/skaffold/run/SkaffoldCommandLineStateTest.kt [hunk 17]
```diff
@@ -141,4 +143,32 @@ class SkaffoldCommandLineStateTest {
...
+            SkaffoldExecutorSettings.ExecutionMode.DEV
+        )
+
+        expectThrows(
+            ExecutionException::class,
+            ThrowableRunnable { skaffoldCommandLineState.startProcess() })
+    }
+
+    @Test
+    fun `A run error isn't thrown if skaffold is in the system PATH`() {
```
Comment (etanshaul) [inline skaffold/src/test/kotlin/com/google/container/tools/skaffold/run/SkaffoldCommandLineStateTest.kt]: I’d remove all the single quotes from these test names. It was causing weird behavior when running the tests from the IDEAs JUnit runner.

**Comment by ZariaHoward** [comment]
screenshot from 2019-02-14 15-36-18

File: skaffold/src/main/resources/messages/SkaffoldBundle.properties [hunk 18]
```diff
@@ -42,3 +42,4 @@ skaffold.override.image.repo.tooltip=<html>The default image repository allows y
 skaffold.no.file.selected.error=Skaffold configuration file is not selected.
 skaffold.invalid.file.error=Skaffold configuration file doesn't exist or invalid.
 skaffold.corrupted.run.settings=Your Skaffold run configuration is corrupted. Please re-create it to fix.
+skaffold.not.on.system.error=Could not detect Skaffold installation. Please install Skaffold. See required dependencies <a href="https://github.com/GoogleContainerTools/google-container-tools-intellij#prerequisites-and-required-dependencies">here</a>:
```
Comment (ivanporty) [inline skaffold/src/main/resources/messages/SkaffoldBundle.properties]: Does this link work in the message? I see it ends with a colon, is this intentional?

File: skaffold/src/test/kotlin/com/google/container/tools/skaffold/DefaultSkaffoldExecutorServiceTest.kt [hunk 19]
```diff
@@ -226,4 +231,18 @@ class DefaultSkaffoldExecutorServiceTest {
...
+
+    @Test
+    fun `isSkaffoldAvailable returns true when skaffold is available`() {
+        mockkStatic(System::class)
+        every { System.getenv("PATH") } answers { testSkaffoldFile.parent }
+        assertThat(defaultSkaffoldExecutorService.isSkaffoldAvailable()).isTrue()
+    }
+
+    @Test
+    fun `isSkaffoldAvailable returns false when skaffold is not available`() {
```
Comment (ivanporty) [inline skaffold/src/test/kotlin/com/google/container/tools/skaffold/DefaultSkaffoldExecutorServiceTest.kt]: It might also be a good idea to add a similar unit test, but with actual list of directories in mocked PATH, not only empty PATH, so we know for sure the parsing works.

**Comment by etanshaul** [comment]
@ZariaHoward turns out the windows test failures are due to the `File#setExecutable`..as tested on my windows vm. I'm not exactly sure why yet.

**Comment by etanshaul** [comment]
actually, it might be the `mockkStatic`

**Comment by etanshaul** [comment]
alternative to using mockkStatic:

in `SkaffoldExecutorService#isSkaffoldAvailable`, extract `System.getenv("PATH")` to a separate function so that it can be mocked. I.e.:

```
fun getSystemPath(): String = System.getenv("PATH")
```

Then in the test, remove all `mockkStatic`, and instead of mocking System.getenv, mock:

`every { defaultSkaffoldExecutorService.getSystemPath() } answers { [whatever you want] }`

File: skaffold/src/main/resources/messages/SkaffoldBundle.properties (hunk 18 above)
Comment (etanshaul) [reply skaffold/src/main/resources/messages/SkaffoldBundle.properties]: It probably should not end with a colon. Also the link does not work. Probably need to attach a URL handler.

Comment (etanshaul) [reply skaffold/src/main/resources/messages/SkaffoldBundle.properties]: I'd suggest, if its not easy to figure out how to get the URL to work here, to just change it from a hyperlink to just print the url (so the user can copy/paste it), and open a separate issue to track getting the link working.

Comment (ivanporty) [reply skaffold/src/main/resources/messages/SkaffoldBundle.properties]: Agree, might be not easy here. We can also think of some sort of notification - they are more flexible.

**Comment by ZariaHoward** [comment]
screenshot from 2019-02-21 15-18-26

**Comment by etanshaul** [comment]
looks good. commit the code updates?

File: common-test-lib/src/main/kotlin/com/google/container/tools/test/ContainerToolsRule.kt [hunk 20]
```diff
@@ -133,6 +133,7 @@ class ContainerToolsRule(private val testInstance: Any) : TestRule {
                 }
             }
 
+            file.setExecutable(true)
```
Comment (ivanporty) [inline common-test-lib/src/main/kotlin/com/google/container/tools/test/ContainerToolsRule.kt]: It does not seem necessary for each test file we are creating, could we add a boolean parameter (optional, defaults to false) to mark file as executable? Or some other way to mark file executable in the unit test itself?

File: skaffold/src/main/kotlin/com/google/container/tools/skaffold/SkaffoldExecutorService.kt [hunk 21]
```diff
@@ -42,9 +48,41 @@ abstract class SkaffoldExecutorService {
...
+            type,
+            NotificationListener.UrlOpeningListener(true)
+        )
+
+
+
     /** Path for Skaffold executable, any form supported by [ProcessBuilder] */
     protected abstract var skaffoldExecutablePath: Path
 
+    open fun getSystemPath(): String = System.getenv("PATH")
```
Comment (etanshaul) [inline skaffold/src/main/kotlin/com/google/container/tools/skaffold/SkaffoldExecutorService.kt]: no need to put this in the interface. you can keep only in the Default impl
